### PR TITLE
Realm2

### PR DESCRIPTION
**Pull request by akashacko**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,7 +13,10 @@ homepage = "https://github.com/ackotech/realm"
 
 [features]
 default = []
-dynamic-elm-dependencies = []
+postgres = ["diesel/postgres"]
```
Comment (amitu) [inline Cargo.toml]: Add in readme that enabling either of postgres+postgres_default or sqlite+sqlite_default features is now required.

File: README.md [hunk 2]
```diff
@@ -17,11 +17,33 @@ Rust / Elm base full stack web framework.
...
+//
+// [build-dependencies]
+// walkdir = "2"
+```
+
+- Elm: Scroll to top of page on page change
+- Elm: Sending a message after delay of 200ms to let app show a loading dialog
+- Elm: BREAKING: removed Realm.Utils.{link, plainLink, newTabLink}, use Element versions
+  instead.
+- Generic ud cookie: Applications can define their own struct for `ud` cookie. This change is not compatible with the previous realm versions. Please Refer to `UDMIGRATION.md` guide.
```
Comment (amitu) [inline README.md]: Make UDMIGRATION.md a link.

File: frontend/Realm.elm [hunk 3]
```diff
@@ -238,6 +242,11 @@ appInit a vflags url key =
     )
 
 
+navigate : String -> Cmd (Msg msg)
+navigate =
+    GoTo >> message
+
+
```
Comment (amitu) [inline frontend/Realm.elm]: Add a note that we have added this method.

File: frontend/Realm/Utils.elm [hunk 4]
```diff
@@ -28,6 +28,16 @@ iff c e =
     yesno c e E.none
 
 
+mif : Maybe a -> (a -> E.Element msg) -> E.Element msg
+mif m f =
```
Comment (amitu) [inline frontend/Realm/Utils.elm]: Add a note that we have added this method.

File: src/base/db.rs [hunk 5]
```diff
@@ -15,70 +10,14 @@ lazy_static! {
...
-        .build(manager)
-        .expect("Failed to create DIESEL_POOL.")
-}
-
-lazy_static! {
-    pub static ref DIESEL_POOLS: antidote::RwLock<HashMap<String, r2d2::Pool<r2d2_diesel::ConnectionManager<RealmConnection>>>> =
-        antidote::RwLock::new(HashMap::new());
-}
-
-pub fn connection() -> r2d2::PooledConnection<r2d2_diesel::ConnectionManager<RealmConnection>> {
```
Comment (amitu) [inline src/base/db.rs]: Add a note that we have moved this and other methods in this file.

File: src/base/in_.rs [hunk 6]
```diff
@@ -1,37 +1,57 @@
...
 
+    pub fn is_anonymous(&self) -> bool {
+        self.ud().is_none()
+    }
+
+    pub fn is_authenticated(&self) -> bool {
+        self.ud().is_some()
+    }
+
+    pub fn is_local(&self) -> bool {
```
Comment (amitu) [inline src/base/in_.rs]: Add notes about these methods.

File: src/base/in_.rs [hunk 7]
```diff
@@ -61,31 +73,30 @@ impl<'a> In<'a> {
         self.get_header(http::header::USER_AGENT)
     }
 
-    pub fn name(&self) -> Option<String> {
-        match *self.ud.borrow() {
-            Some((_, ref name, _)) => Some(name.to_string()),
-            None => None,
-        }
+    pub fn reset_ud(&self) {
```
Comment (amitu) [inline src/base/in_.rs]: Do we need both reset_ud and logout?

File: src/lib.rs [hunk 8]
```diff
@@ -77,6 +93,13 @@ pub enum Error {
     },
 }
 
+pub fn error<T>(key: &str, message: &str) -> std::result::Result<T, failure::Error> {
```
Comment (amitu) [inline src/lib.rs]: Add a note that we have this method now.

File: src/request_config.rs [hunk 9]
```diff
@@ -296,6 +296,29 @@ impl RequestConfig {
         })
     }
 
+    pub fn json<T>(&mut self, name: &str) -> Result<T, Error>
```
Comment (amitu) [inline src/request_config.rs]: Add a note about this method.

File: src/serve.rs [hunk 10]
```diff
@@ -23,24 +30,50 @@ macro_rules! realm_serve {
...
+            match res {
+                Ok(r) => r.into_inner().unwrap(),
+                Err(_) => {
+                    // https://docs.rs/log-panics/2.0.0/src/log_panics/lib.rs.html#50-87
+                    // we are not getting any meaningful message to print here. one way
+                    // to do it would be set a hook as described in link above, and
+                    // since hooks are global, store the panic message in a global
+                    // hashmap (thread id-message), and retrieve the message by sending
+                    // the current thread id.
+                    Ok(realm::serve::server_error("panic".to_string().into_bytes()))
```
Comment (amitu) [inline src/serve.rs]: Add a note that we show backtrace for error now (panic are still not clean).

File: src/base/in_.rs (hunk 7 above)
Comment (akashacko) [reply src/base/in_.rs]: No. Both are doing the same thing.

**Comment by amitu** [comment]
@AbrarNitk: please review this PR and use it with teja to make sure everything is fine (as per README). We want to merge it asap now.

**Pull request merged**